lib/scanner/xml-scanner/xml-scanner.c

```diff
@@ -0,0 +1,383 @@
+                                    };
+  ScratchBuffersMarker marker;
+  scratch_buffers_mark(&marker);
+  self->xml_ctx = g_markup_parse_context_new(&scanner_callbacks, 0, self, NULL);
+  g_markup_parse_context_parse(self->xml_ctx, input, input_len, error);
+  if (error && *error)
+    goto exit;
+  g_markup_parse_context_end_parse(self->xml_ctx, error);
+
+exit:
+  scratch_buffers_reclaim_marked(marker);
+  g_markup_parse_context_free(self->xml_ctx);
+  self->xml_ctx = NULL;
+}
+
+
+/*
+  For some patterns, GPatternSpec stores the pattern as reversed
+  string. In such cases, the matching must be executed against the
+  reversed matchstring. If one uses g_pattern_match_string, glib will
+  reverse the string internally in these cases, but it is suboptimal if
+  the same matchstring must be matched against different patterns,
+  because memory is allocated each time and string is copied as
+  reversed, though it would be enough to execute this reverse once. For
+  that reason one can use g_pattern_match(), which accepts both the
+  matchstring and a reversed matchstring as parameters.
+
+  Though there are cases when no reverse is needed at all. This is
+  decided in the constructor of GPatternSpec, but at this time of
+  writing this information is not exported, and cannot be extracted in
+  any safe way, because GPatternSpec is forward declared.
+
+  This function below is an oversimplified/safe version of the logic
+  used glib to decide if the pattern will be reversed or not. One need
+  to note the worst case scenario is if syslog-ng will not reverse the
+  matchstring, though it should because in that case number of extra
+  memory allocations and string reversals scale linearly with the number
+  of patterns. We need to avoid not pre-reversing, when glib would.
+ */
+gboolean
+joker_or_wildcard(GList *patterns)
+{
+  gboolean retval = FALSE;
+  GList *pattern = patterns;
+  while (pattern != NULL)
+    {
+      gchar *str = ((gchar *)pattern->data);
+      if (strpbrk(str, "*?"))
+        {
+          retval = TRUE;
+          break;
+        }
+      pattern = g_list_next(pattern);
+    }
+
+  return retval;
+}
+
+static void
+_compile_and_add(gpointer tag_glob, gpointer exclude_patterns)
+{
+  g_ptr_array_add(exclude_patterns, g_pattern_spec_new(tag_glob));
+}
+
+void
+xml_scanner_init(XMLScanner *self, XMLScannerOptions *options, PushCurrentKeyValueCB push_function,
+                 gpointer user_data, gchar *key_prefix)
+{
+  memset(self, 0, sizeof(*self));
+  self->start_element_function = xml_scanner_start_element_method;
+  self->end_element_function = xml_scanner_end_element_method;
+  self->text_function = xml_scanner_text_method;
+  self->should_push_text = xml_scanner_should_push_text_method;
+
+  self->options = options;
+  self->exclude_patterns = g_ptr_array_new_with_free_func((GDestroyNotify)g_pattern_spec_free);
+  g_list_foreach(self->options->exclude_tags, _compile_and_add, self->exclude_patterns);
```

Comment:
Good point.

Reply:
I have given some thought to this idea and I don't think we should change this:
- you are right about that this happens per message now, but only because each parser in syslog-ng creates the scanner on the stack in the `process()` function.
This is made to avoid problems of cloning pipes I think?
- If we would do these initializations in a setter function, it wouldn't really be just a setter function.
I would like to have another reviewer's opinion before doing this change.
Besides `exclude_patterns` is an attribute of XMLScanner.
Attribute initialization of an object should be done in a constructor, right?

Maybe:
XMLScannerOptions should contain only the result: the `exclude_pattens`, and also the `matchstring_shouldreverse`. `exclude_tags` should not be stored in the options.

Reply:
After our discussion IRL we concluded that the `xml_scanner_options_set_exclude_tags` can do the compile too, but then the setter function should be renamed to show its more complex functionality.